File: frontend/src/api/githubApp.ts
/**
 * GitHub App + Device Flow client (ADR-0035, IMPL-0010).
 *
 * Mirrors the four backend endpoints under /api/integrations/github.
 * The frontend never talks to GitHub directly — it polls our backend's
 * /status endpoint while our backend polls GitHub's token endpoint.
 */

import { useEffect, useRef, useState } from 'react'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { request } from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type DeviceFlowPollingStatus =
  | 'installation_pending'
  | 'device_pending'
  | 'connected'
  | 'expired'
  | 'denied'
  | 'rate_limited'
  | 'error'

export interface DeviceFlowConnectResponse {
  user_code: string
  verification_uri: string
  expires_in: number
  interval: number
  install_url: string
}

export interface DeviceFlowStatusResponse {
  status: DeviceFlowPollingStatus
  user_code: string | null
  expires_at: string | null
  installation_id: number | null
  github_login: string | null
  error: string | null
}

export interface DeviceFlowDisconnectResponse {
  status: 'disconnected'
  manual_revoke_url: string
}

export interface PushAccessDiagnoseResponse {
  can_push: boolean
  reason: string
  repo_url: string
  /** ISO 8601 UTC. Backend echoes the *probe* time, not the response
   *  time — so a cached result still shows "checked 2 min ago". */
  checked_at: string
}

// ---------------------------------------------------------------------------
// Raw API
// ---------------------------------------------------------------------------

export const githubAppApi = {
  /**
   * Initiate (or resume) the device flow.
   * Pass *returnTo* to make the post-install redirect target a specific
   * SPA path (e.g. '/onboarding/connect'). Defaults to /settings.
   */
  connect: (opts?: { returnTo?: string }) => {
    const qs = opts?.returnTo
      ? `?return_to=${encodeURIComponent(opts.returnTo)}`
      : ''
    return request<DeviceFlowConnectResponse>(
      `/api/integrations/github/connect${qs}`,
      { method: 'POST', body: '{}' },
    )
  },
  status: () =>
    request<DeviceFlowStatusResponse>('/api/integrations/github/status'),
  /** Force an immediate poll tick on the backend and return the
   * resulting status. Used after the user clicks Authorize on
   * github.com so we don't wait for the next scheduled poll
   * (which can be 5-60s out depending on GitHub's slow_down). */
  pollNow: () =>
    request<DeviceFlowStatusResponse>(
      '/api/integrations/github/poll-now',
      { method: 'POST', body: '{}' },
    ),
  disconnect: () =>
    request<DeviceFlowDisconnectResponse>(
      '/api/integrations/github/disconnect',
      { method: 'POST', body: '{}' },
    ),
  /** Probe push access against the currently-configured GitHub repo.
   *  Backend caches for 5 minutes; pass ``refresh: true`` to force a
   *  fresh GitHub call after the user fixes the App on github.com. */
  diagnose: (opts?: { refresh?: boolean }) => {
    const qs = opts?.refresh ? '?refresh=1' : ''
    return request<PushAccessDiagnoseResponse>(
      `/api/integrations/github/diagnose${qs}`,
    )
  },
}

// ---------------------------------------------------------------------------
// React Query hooks
// ---------------------------------------------------------------------------

export function useGithubAppConnect() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (opts?: { returnTo?: string }) => githubAppApi.connect(opts),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['github-app', 'status'] })
    },
  })
}

/**
 * Polls /status while *enabled* is true. The backend reports terminal
 * states ('connected' | 'expired' | 'denied' | 'error') — the caller
 * disables polling once one of those arrives.
 *
 * 404 means no flow has ever been started; we surface that as
 * ``status: null`` so the UI renders the "Connect GitHub" button.
 */
export function useGithubAppStatus(opts: {
  enabled?: boolean
  intervalMs?: number
}) {
  const { enabled = false, intervalMs = 2000 } = opts
  return useQuery({
    queryKey: ['github-app', 'status'],
    queryFn: async (): Promise<DeviceFlowStatusResponse | null> => {
      try {
        return await githubAppApi.status()
      } catch (err) {
        if (err instanceof Error && err.message.startsWith('404:')) {
          return null
        }
        throw err
      }
    },
    enabled,
    refetchInterval: (query) => {
      const data = query.state.data as DeviceFlowStatusResponse | null
      if (!data) return false
      const terminal: DeviceFlowPollingStatus[] = [
        'connected',
        'expired',
        'denied',
        'error',
      ]
      return terminal.includes(data.status) ? false : intervalMs
    },
  })
}

export function useGithubAppPollNow() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () => githubAppApi.pollNow(),
    onSuccess: (data) => {
      // Push the poll result straight into the cache so the modal's
      // useGithubAppStatus picks it up without waiting for its next
      // refetch interval.
      qc.setQueryData(['github-app', 'status'], data)
    },
  })
}


/**
 * Probe push access for the configured GitHub repo (IMPL-0018 / B35c).
 *
 * Returns ``data: null`` on 404 so the badge can render nothing without
 * a thrown error bubbling through React Query's ``isError`` state. Every
 * other failure mode bubbles as a normal query error so the operator
 * gets a visible signal instead of a phantom "everything looks fine".
 *
 * Stale window matches the backend cache (5 minutes) so the SPA and the
 * server agree on what "fresh" means. Background refetch on focus is
 * disabled — the user comes back to the Settings page after fixing the
 * App on github.com, and the explicit Refresh button is the natural
 * way to re-probe.
 */
export function useGitHubPushDiagnose() {
  return useQuery({
    queryKey: ['github-app', 'diagnose'],
    queryFn: async (): Promise<PushAccessDiagnoseResponse | null> => {
      try {
        return await githubAppApi.diagnose()
      } catch (err) {
        // 404 = GitHub not configured. The badge renders nothing in
        // this state — don't surface a query error to the caller, since
        // the absence of a GitHub integration is normal pre-onboarding.
        if (err instanceof Error && err.message.startsWith('404:')) {
          return null
        }
        throw err
      }
    },
    // 5 min — matches the server-side cache so we don't re-fetch faster
    // than the backend is willing to re-probe.
    staleTime: 5 * 60 * 1000,
    refetchOnWindowFocus: false,
  })
}

export function useGithubAppDisconnect() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () => githubAppApi.disconnect(),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['github-app', 'status'] })
      qc.invalidateQueries({ queryKey: ['integrations'] })
    },
  })
}

/**
 * Decides whether to open the device-flow modal on page mount.
 *
 * Triggers ONLY on the ``?github_setup=complete`` or
 * ``?github_setup=updated`` query parameter — i.e. the user just came
 * back from the backend's /setup callback after a successful install
 * or configure-save on github.com. In both cases the modal opens so
 * the user can finish authorising the device.
 *
 * We deliberately do NOT auto-open the modal whenever the backend
 * has an in-flight row, because that fires PRE-install too (immediately
 * after the user clicks Install but before they've actually done
 * anything on GitHub). The "App already installed → GitHub diverts to
 * Configure → setup_url never fires" recovery case is handled by an
 * explicit "Resume install" affordance on the page rather than a
 * surprise modal — see ``manualResume`` in ConnectRepo / IntegrationSettings.
 *
 * Lives at this layer (not inside the connect button) so the post-
 * install resume flow works regardless of whether any specific CTA is
 * currently mounted - once the install creates an integration row the
 * catalog button unmounts, but the page still renders.
 */
export function useGithubAppResumeOnReturn(): {
  response: DeviceFlowConnectResponse | null
  clear: () => void
  /** Manually open the modal (e.g. from a "Resume install" button).
   *  Calls the idempotent /connect to fetch the in-flight code and
   *  mounts the modal in the same way the URL-driven path does. */
  resume: () => Promise<void>
} {
  const connect = useGithubAppConnect()
  const [response, setResponse] = useState<DeviceFlowConnectResponse | null>(
    null,
  )
  const triggered = useRef(false)

  useEffect(() => {
    if (typeof window === 'undefined') return
    if (triggered.current) return
    if (connect.isPending || response) return

    const url = new URL(window.location.href)
    const setupStatus = url.searchParams.get('github_setup')
    if (setupStatus !== 'complete' && setupStatus !== 'updated') return

    triggered.current = true

    const cleanUrl = () => {
      url.searchParams.delete('github_setup')
      url.searchParams.delete('integration_id')
      url.searchParams.delete('reason')
      // ``url.toString()`` already serialises the hash from the original
      // URL — appending ``window.location.hash`` would double it
      // (#integrations#integrations).
      window.history.replaceState({}, '', url.toString())
    }

    void (async () => {
      try {
        const r = await connect.mutateAsync({})
        setResponse(r)
      } finally {
        cleanUrl()
      }
    })()
    // Fire once on mount.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const resume = async () => {
    if (connect.isPending || response) return
    const r = await connect.mutateAsync({})
    setResponse(r)
  }

  return {
    response,
    clear: () => {
      setResponse(null)
      // Allow a fresh trigger if the user starts a new flow later.
      triggered.current = false
    },
    resume,
  }
}

File: backend/opensec/api/routes/github_app.py
"""GitHub App + Device Flow onboarding routes (ADR-0035, IMPL-0010).

Mounted at ``/api/integrations/github``. Four endpoints:

- ``POST /connect``    — kick off (or resume) the device flow.
- ``GET  /setup``      — landing page after Install on github.com.
- ``GET  /status``     — frontend polls this every ~2s while modal open.
- ``POST /disconnect`` — drop credentials + installation row locally.

The user access token issued by the device flow lands in the existing
credential vault under the same key the MCP Gateway already substitutes,
so workspaces transparently keep working without any agent-side change.
"""

from __future__ import annotations

import asyncio
import logging
import time
from typing import TYPE_CHECKING, Literal
from urllib.parse import quote, urlencode

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from fastapi.responses import RedirectResponse

from opensec.config import settings
from opensec.db import repo_integration
from opensec.db.connection import get_db
from opensec.integrations.github_app import repo as gh_repo
from opensec.integrations.github_app.client import GitHubDeviceFlowClient
from opensec.integrations.github_app.flow import (
    DeviceFlowOrchestrator,
    InstallationCsrfMismatchError,
)
from opensec.integrations.github_app.models import (
    DeviceFlowConnectResponse,
    DeviceFlowDisconnectResponse,
    DeviceFlowStatusResponse,
)
from opensec.models import IntegrationConfigCreate

if TYPE_CHECKING:
    import aiosqlite

    from opensec.integrations.audit import AuditLogger
    from opensec.integrations.github_app.flow import GithubAppClientProtocol
    from opensec.integrations.vault import CredentialVault

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/integrations/github", tags=["github-app"])

# These MUST match the constants used by routes/onboarding.py
# (GITHUB_ADAPTER_TYPE, GITHUB_PROVIDER_NAME). Every downstream consumer
# (_engine_dep.py, workspaces.py, executor.py, connection_tester.py)
# looks up the GitHub integration by these exact values; a mismatch
# would silently make the assessment + workspace-clone paths run
# unauthenticated even though the device flow succeeded.
GITHUB_PROVIDER = "GitHub"
GITHUB_ADAPTER = "github"
GITHUB_MANUAL_REVOKE_URL = "https://github.com/settings/applications"
DEFAULT_RETURN_PATH = "/settings"

# Allow-list for return_to paths we'll redirect to after a successful App
# install. Any value POSTed to /connect must match one of these prefixes
# verbatim — otherwise it's silently dropped and the default is used.
# Prevents an attacker from coercing the redirect to an arbitrary site.
_ALLOWED_RETURN_PATHS: tuple[str, ...] = (
    "/settings",
    "/onboarding/connect",
)


def _sanitize_return_path(value: str | None) -> str:
    """Validate that *value* is a known onboarding/settings path."""
    if not value:
        return DEFAULT_RETURN_PATH
    if not value.startswith("/"):
        return DEFAULT_RETURN_PATH
    for prefix in _ALLOWED_RETURN_PATHS:
        if value == prefix or value.startswith(prefix + "?") or value.startswith(prefix + "#"):
            return value
    return DEFAULT_RETURN_PATH


# In-memory map csrf_state -> (return_path, expires_at). Process-lifetime.
# Acceptable because the device-flow window is at most 15 minutes; if
# the process restarts mid-flow the user lands on the default path,
# which is still a valid recovery point. Keeps the schema migration-free.
#
# Each entry is bounded by ``RETURN_PATH_TTL_SECONDS`` so an abandoned
# /connect (user never installs) doesn't leak memory under repeated
# clicks. ``MAX_RETURN_PATHS`` is a hard cap on entry count as a
# secondary defense against pathological abuse.
RETURN_PATH_TTL_SECONDS = 30 * 60  # comfortably above the 15-minute device-code window
MAX_RETURN_PATHS = 256


def _return_paths(request: Request) -> dict[str, tuple[str, float]]:
    state = request.app.state
    existing = getattr(state, "github_app_return_paths", None)
    if existing is None:
        existing = {}
        state.github_app_return_paths = existing
    return existing


def _evict_expired_return_paths(paths: dict[str, tuple[str, float]]) -> None:
    """Drop entries past their TTL. Called on every read/write so the
    map self-heals without a separate sweeper task."""
    now = time.time()
    fresh = {k: v for k, v in paths.items() if v[1] > now}
    if len(fresh) != len(paths):
        paths.clear()
        paths.update(fresh)


def _stash_return_path(request: Request, csrf_state: str, path: str) -> None:
    paths = _return_paths(request)
    _evict_expired_return_paths(paths)
    # Hard cap (defense in depth): drop the oldest entries if we're at
    # the cap. Wouldn't trigger under normal use but stops a malicious
    # client from blowing memory by spamming /connect.
    if len(paths) >= MAX_RETURN_PATHS:
        # Drop the oldest-expiring entries to make room.
        ordered = sorted(paths, key=lambda k: paths[k][1])
        drop_count = max(1, len(paths) - MAX_RETURN_PATHS + 1)
        for key in ordered[:drop_count]:
            paths.pop(key, None)
    paths[csrf_state] = (path, time.time() + RETURN_PATH_TTL_SECONDS)


def _pop_return_path(request: Request, csrf_state: str) -> str:
    paths = _return_paths(request)
    _evict_expired_return_paths(paths)
    entry = paths.pop(csrf_state, None)
    if entry is None:
        return DEFAULT_RETURN_PATH
    path, _ = entry
    return path


# ---------------------------------------------------------------------------
# Construction helpers (overridable in tests)
# ---------------------------------------------------------------------------


def _build_github_client() -> GithubAppClientProtocol:
    """Build a real GitHubDeviceFlowClient. Tests override this via patch."""
    return GitHubDeviceFlowClient(client_id=settings.github_app_client_id)


def _build_orchestrator(
    db: aiosqlite.Connection,
    vault: CredentialVault,
    audit: AuditLogger,
) -> DeviceFlowOrchestrator:
    return DeviceFlowOrchestrator(
        db=db,
        vault=vault,
        audit=audit,
        client_factory=_build_github_client,
        app_slug=settings.github_app_slug,
        client_id=settings.github_app_client_id,
    )


def _get_orchestrator(
    request: Request,
    db: aiosqlite.Connection,
    vault: CredentialVault,
    audit: AuditLogger,
) -> DeviceFlowOrchestrator:
    """Process-singleton orchestrator stored on ``app.state``.

    Why a singleton: the orchestrator owns an in-memory map of running
    polling tasks keyed by integration_id. If we built a fresh
    orchestrator per request, ``start()`` would spawn a new task on
    every /connect (orphaning the previous one) and ``disconnect``
    would never find a task to cancel. One instance per process keeps
    the lifecycle clean — and lets the FakeClient transport survive
    across requests in tests too.
    """
    existing = getattr(request.app.state, "github_app_orchestrator", None)
    if existing is not None:
        return existing
    orchestrator = _build_orchestrator(db, vault, audit)
    request.app.state.github_app_orchestrator = orchestrator
    return orchestrator


def _require_app_configured() -> None:
    if not settings.github_app_client_id or not settings.github_app_slug:
        raise HTTPException(
            status_code=503,
            detail=(
                "GitHub App onboarding is not configured on this instance. "
                "Set OPENSEC_GITHUB_APP_CLIENT_ID and OPENSEC_GITHUB_APP_SLUG, "
                "or fall back to the personal access token form."
            ),
        )


def _require_vault_and_audit(request: Request) -> tuple[CredentialVault, AuditLogger]:
    vault = getattr(request.app.state, "vault", None)
    audit = getattr(request.app.state, "audit_logger", None)
    if vault is None or audit is None:
        raise HTTPException(
            status_code=503,
            detail="Credential vault or audit logger unavailable.",
        )
    return vault, audit


def _install_url(csrf_state: str) -> str:
    # Quote the slug too — the value comes from env (trusted by intent),
    # but defending against a typo with ``/`` or ``?`` in it is free.
    slug = quote(settings.github_app_slug, safe="")
    return (
        f"https://github.com/apps/{slug}"
        f"/installations/new?state={quote(csrf_state, safe='')}"
    )


def _resolve_frontend_base_url() -> str:
    """Pick the right origin for the post-install redirect.

    Priority:
    1. Explicit override via ``OPENSEC_FRONTEND_BASE_URL``.
    2. ``OPENSEC_STATIC_DIR`` set → backend serves the SPA on the same
       origin as the API → use ``base_url``.
    3. Neither set → assume Vite dev convention on ``:5173``.
    """
    if settings.frontend_base_url:
        return settings.frontend_base_url.rstrip("/")
    if settings.static_dir:
        return settings.base_url.rstrip("/")
    return "http://localhost:5173"


def _frontend_redirect(return_path: str, *, status: str, **extra: str) -> str:
    """Build the post-callback redirect back to the SPA.

    *return_path* is a sanitized SPA-relative path (e.g. ``/settings``
    or ``/onboarding/connect``). Anchors and query params on the path
    are preserved when present; otherwise we add the section anchor
    for /settings so the browser scrolls to the integrations section.
    The ``?github_setup=...`` query param is what
    ``useGithubAppResumeOnReturn`` reads on mount to re-open the modal.
    """
    params = {"github_setup": status, **extra}
    base = _resolve_frontend_base_url()

    # Split off any pre-existing query/hash so we layer ours on top
    # without colliding.
    path_only = return_path
    existing_hash = ""
    if "#" in path_only:
        path_only, existing_hash = path_only.split("#", 1)
        existing_hash = f"#{existing_hash}"
    if "?" in path_only:
        # Drop the user-supplied query — our params are the only ones
        # the resume hook cares about.
        path_only = path_only.split("?", 1)[0]

    # Default scroll anchor for the settings page (where the
    # integrations section lives). Onboarding pages keep no anchor.
    if not existing_hash and path_only == "/settings":
        existing_hash = "#integrations"

    return f"{base}{path_only}?{urlencode(params)}{existing_hash}"


# ---------------------------------------------------------------------------
# Singleton-integration helper
# ---------------------------------------------------------------------------


def _connect_lock(request: Request) -> asyncio.Lock:
    """Process-singleton lock guarding _resolve_or_create_app_integration.

    Two concurrent /connect calls (rapid double-click) would otherwise
    both miss the SELECT-then-INSERT check and create duplicate
    integration_config rows for ``provider_name='GitHub'``. Single-user
    mode is single-process, so an asyncio.Lock is sufficient. (For
    SaaS multi-user we'd need a DB-level UNIQUE partial index — out
    of scope for this PR.)
    """
    state = request.app.state
    existing = getattr(state, "github_app_connect_lock", None)
    if existing is None:
        existing = asyncio.Lock()
        state.github_app_connect_lock = existing
    return existing


async def _resolve_or_create_app_integration(db: aiosqlite.Connection) -> str:
    """Find the singleton App-onboarded github integration, creating one if needed.

    The PAT integration (if any) lives as a separate row and is left
    untouched until a successful App connect archives it via
    DeviceFlowOrchestrator.

    Caller must hold ``_connect_lock(request)`` to make the
    SELECT-then-INSERT atomic across concurrent /connect calls.
    """
    cursor = await db.execute(
        """
        SELECT integration_config.id
        FROM integration_config
        INNER JOIN github_app_installation
                ON github_app_installation.integration_id = integration_config.id
        WHERE integration_config.provider_name = ?
        LIMIT 1
        """,
        (GITHUB_PROVIDER,),
    )
    row = await cursor.fetchone()
    if row is not None:
        return row["id"]

    integ = await repo_integration.create_integration(
        db,
        IntegrationConfigCreate(
            adapter_type=GITHUB_ADAPTER,
            provider_name=GITHUB_PROVIDER,
            enabled=False,
            config=None,
            action_tier=0,
        ),
    )
    return integ.id


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------


@router.post("/connect", response_model=DeviceFlowConnectResponse)
async def connect(
    request: Request,
    return_to: str | None = Query(default=None),
    db: aiosqlite.Connection = Depends(get_db),
) -> DeviceFlowConnectResponse:
    _require_app_configured()
    vault, audit = _require_vault_and_audit(request)

    async with _connect_lock(request):
        integration_id = await _resolve_or_create_app_integration(db)
        orchestrator = _get_orchestrator(request, db, vault, audit)
        started = await orchestrator.initiate(integration_id)
    # Spawn (or re-attach to) the background polling loop so the access
    # token is fetched as soon as the user authorizes on github.com —
    # without this the /status endpoint would never advance off
    # device_pending. Idempotent.
    await orchestrator.start(integration_id)

    # Stash a sanitized return path keyed by CSRF state. The setup
    # callback reads it post-install and redirects there instead of the
    # default /settings, so the App flow seamlessly returns the user to
    # whichever page (onboarding, settings) started the flow.
    _stash_return_path(request, started.csrf_state, _sanitize_return_path(return_to))

    return DeviceFlowConnectResponse(
        user_code=started.user_code,
        verification_uri=started.verification_uri,
        expires_in=started.expires_in,
        interval=started.interval,
        install_url=_install_url(started.csrf_state),
    )


@router.get("/setup")
async def setup_callback(
    request: Request,
    state: str = Query(...),
    installation_id: int = Query(...),
    # GitHub sends ``setup_action`` ∈ {"install", "update"}. ``install``
    # is the first time the user installs OpenSec on the account;
    # ``update`` fires when they revisit the App page and reconfigure
    # repos. We honour both, but tag the redirect so the SPA can show
    # different copy ("Connected" vs "Configuration updated").
    setup_action: Literal["install", "update"] = Query("install"),
    db: aiosqlite.Connection = Depends(get_db),
) -> RedirectResponse:
    _require_app_configured()
    vault, audit = _require_vault_and_audit(request)
    orchestrator = _get_orchestrator(request, db, vault, audit)
    return_path = _pop_return_path(request, state)

    redirect_status = "updated" if setup_action == "update" else "complete"

    try:
        await orchestrator.attach_installation(
            csrf_state=state, installation_id=installation_id
        )
    except InstallationCsrfMismatchError:
        return RedirectResponse(
            _frontend_redirect(return_path, status="error", reason="csrf"),
            status_code=302,
        )
    return RedirectResponse(
        _frontend_redirect(return_path, status=redirect_status),
        status_code=302,
    )


@router.get("/status", response_model=DeviceFlowStatusResponse)
async def status(
    request: Request, db: aiosqlite.Connection = Depends(get_db)
) -> DeviceFlowStatusResponse:
    _require_app_configured()
    # Find any github integration that has an installation row (in-flight or
    # connected). If none exists, 404 — caller should POST /connect.
    cursor = await db.execute(
        """
        SELECT integration_id FROM github_app_installation
        ORDER BY updated_at DESC LIMIT 1
        """,
    )
    row = await cursor.fetchone()
    if row is None:
        raise HTTPException(status_code=404, detail="No GitHub App flow in progress")

    integration_id = row["integration_id"]
    record = await gh_repo.get_for_integration(db, integration_id)
    assert record is not None
    return DeviceFlowStatusResponse(
        status=record.polling_status,
        user_code=record.user_code,
        expires_at=record.device_code_expires_at,
        installation_id=record.installation_id,
        github_login=record.github_login,
        error=record.polling_error,
    )


@router.post("/poll-now", response_model=DeviceFlowStatusResponse)
async def poll_now(
    request: Request, db: aiosqlite.Connection = Depends(get_db)
) -> DeviceFlowStatusResponse:
    """Force an immediate poll tick + return the resulting status.

    The background polling loop honors GitHub's stored interval (5s
    minimum, often higher after a ``slow_down``). When the user clicks
    Authorize and comes back to OpenSec, they shouldn't have to wait
    for the next scheduled tick — the SPA hits this endpoint on the
    visibility-change event so the modal flips to Connected within
    one round-trip instead of up to a minute.
    """
    _require_app_configured()
    vault, audit = _require_vault_and_audit(request)

    cursor = await db.execute(
        """
        SELECT integration_id FROM github_app_installation
        ORDER BY updated_at DESC LIMIT 1
        """
    )
    row = await cursor.fetchone()
    if row is None:
        raise HTTPException(status_code=404, detail="No GitHub App flow in progress")

    integration_id = row["integration_id"]
    orchestrator = _get_orchestrator(request, db, vault, audit)
    # Run a single poll step out-of-band. Idempotent on terminal rows.
    await orchestrator.run_poll_step(integration_id)

    record = await gh_repo.get_for_integration(db, integration_id)
    assert record is not None
    return DeviceFlowStatusResponse(
        status=record.polling_status,
        user_code=record.user_code,
        expires_at=record.device_code_expires_at,
        installation_id=record.installation_id,
        github_login=record.github_login,
        error=record.polling_error,
    )


@router.post("/disconnect", response_model=DeviceFlowDisconnectResponse)
async def disconnect(
    request: Request, db: aiosqlite.Connection = Depends(get_db)
) -> DeviceFlowDisconnectResponse:
    # Deliberately NOT gated on _require_app_configured: if the operator
    # unsets OPENSEC_GITHUB_APP_CLIENT_ID after a user has connected, we
    # still want the cleanup path to work so they aren't stranded with a
    # non-refreshable token in the vault.
    vault, audit = _require_vault_and_audit(request)

    # Deterministic ordering: prefer the most-recently-connected row, then
    # most-recently-touched. With the singleton invariant only one row
    # exists in single-user mode; the ordering is defense in depth for
    # forward-compatibility with multi-install SaaS.
    cursor = await db.execute(
        """
        SELECT integration_id FROM github_app_installation
        ORDER BY connected_at DESC, updated_at DESC
        LIMIT 1
        """
    )
    row = await cursor.fetchone()
    if row is not None:
        orchestrator = _get_orchestrator(request, db, vault, audit)
        await orchestrator.disconnect(row["integration_id"])

    return DeviceFlowDisconnectResponse(
        manual_revoke_url=GITHUB_MANUAL_REVOKE_URL,
    )


# ---------------------------------------------------------------------------
# Test seam — drives a single polling tick deterministically
# ---------------------------------------------------------------------------


async def _tick_poll_for_test() -> None:
    """Run one poll step on every in-flight installation.

    Used by route tests to exercise transitions without spinning up the
    background polling task. Production code uses the orchestrator's
    ``start()`` task instead.
    """
    from opensec.db import connection as db_connection
    from opensec.main import app as fastapi_app

    db = db_connection._db
    if db is None:
        return
    vault = getattr(fastapi_app.state, "vault", None)
    audit = getattr(fastapi_app.state, "audit_logger", None)
    if vault is None or audit is None:
        return

    cursor = await db.execute(
        "SELECT integration_id FROM github_app_installation"
    )
    for row in await cursor.fetchall():
        orchestrator = getattr(fastapi_app.state, "github_app_orchestrator", None)
        if orchestrator is None:
            orchestrator = _build_orchestrator(db, vault, audit)
            fastapi_app.state.github_app_orchestrator = orchestrator
        await orchestrator.run_poll_step(row["integration_id"])
